doc/man/src/ocpidev2.1.txt
OCPIDEV2(1)
===========


NAME
----
ocpidev2 - command-line tool for managing OpenCPI assets


SYNOPSIS
--------
*`ocpidev2`* ['<options>'] '<verb>' ['<noun>' ['<name>' ]]


DESCRIPTION
-----------
The *`ocpidev2(1)`* command-line tool is used to perform various
development-related tasks as well as retrieve information
about the environment.
The *`ocpidev2`* command
has full tab completion for its options and arguments.

For convenience, options can occur anywhere in the command.
The general usage concept is: perform the '<verb>' operation on the
'<noun>' asset type whose name is '<name>'.

Man pages are available for *`ocpidev2`* verbs and nouns by typing
the command *`man ocpidev-`*'<verb>' or *`man ocpidev2-`*'<noun>'.


VERBS
-----
The verbs supported by *`ocpidev2`* are:

*`build`*::
    Build the asset(s), running appropriate tools to create
    the binary files and viewable documentation.

*`clean`*::
    Remove all the generated and compiled files.

*`create`*::
    Create the named asset, creating files and directories as required,
    and creating any skeleton files for future editing.

*`show`*::
    Display information about assets.
    
*`register`*, *`unregister`*::
    Register/unregister a project.
    
NOUNS
-----
For supported nouns, see the verb-specific man page available by typing
the command *`man ocpidev2-`*'<verb>' or *`man ocpidev2-`*'<noun>'.


OPTIONS
-------
Options are either single letters following one hyphen or complete
words or acronyms following two hyphens and separated by hyphens.
In option descriptions, a plus sign (+) after the option indicates
that it can be specified more than once on the command line.
Arguments to word-form options are preceded by an equal sign (=)
(as shown in the descriptions) or a space. Arguments to
letter options are preceded by a space.

Most options are only valid for specific verbs or nouns and are
described in the individual verb and noun man pages.

See the link:opencpi.1.html[opencpi(1)] man page
for the general-purpose options that apply to all OpenCPI tools.

BUGS
----
See https://www.opencpi.org/report-defects


RESOURCES
---------
See the main web site: https://www.opencpi.org

See the 'OpenCPI Component Development Guide':
https://opencpi.gitlab.io/releases/latest/docs/OpenCPI_Component_Development_Guide.pdf

See the 'OpenCPI User Guide':
https://opencpi.gitlab.io/releases/latest/docs/OpenCPI_User_Guide.pdf


SEE ALSO
--------
link:ocpidev2-build.1.html[ocpidev2-build(1)]
link:ocpidev2-clean.1.html[ocpidev2-clean(1)]
link:ocpidev2-create.1.html[ocpidev2-create(1)]
link:ocpidev2-register.1.html[ocpidev2-register(1)]
link:ocpidev2-show.1.html[ocpidev2-show(1)]
link:ocpidev2-unregister.1.html[ocpidev2-unregister(1)]

COPYING
-------
Copyright \(C) 2023 OpenCPI www.opencpi.org. OpenCPI is free software:
you can redistribute it and/or modify it under the terms of the
GNU Lesser General Public License as published by the Free
Software Foundation, either version 3 of the License, or (at your
option) any later version.


CHANGELOG.md
# [v2.4.7](https://gitlab.com/opencpi/opencpi/-/compare/v2.4.6...v2.4.7) (2024-01-04)

Changes/additions since [OpenCPI Release v2.4.6](https://gitlab.com/opencpi/opencpi/-/releases/v2.4.6)

### Features
- **app**: updated the fsk_dig_radio_ctrl image. (!1057)(b6fcacfb)
- **comp**: support include and xi:include. (!995)(ebd35042)
- **comp**: replace xi:include in property files with include. (!995)(ebd35042)
- **comp,tests**: add ability to compare top N bits of fractional time in ocpi_testing. (!1144)(45d4d09e)
- **devops**: enable autoscaling for CI/CD pipelines. (!1115)(2b0e5ab6)
- **devops**: enable parallel building of workers in CI/CD pipelines. (!1115)(2b0e5ab6)
- **devops**: allow for skipping of certain CI/CD pipeline jobs if relevant changes are not made to the repo. (!1120)(5c7c286a)
- **devops**: make non-MR CI/CD pipelines start automatically after being launched. (!1124)(2da5d1c8)
- **devops**: remove sudo from centos7 docker images. (!1125)(048f09e4)
- **devops**: update ubuntu dockerfiles. (!1126)(8f2019c3)
- **devops**: update rocky8 dockerfile. (!1128)(6b9a432d)
- **devops**: allow projects using downstream-ci.yml to pass the `CI_OCPI_PLATFORMS` and `CI_OCPI_PROJECTS` variables when set in the project settings. (!1130)(baad9f06)
- **devops**: re-implement CI/CD pipeline HWIL. (!1137)(92efb17b)
- **devops**: enabling CI/CD pipelines to deploy full docker images to public registry. (!1141)(8c4d366d)
- **devops**: add Ubuntu18.04 support to CI pipelines. (!842)(2fc0efd5)
- **devops**: don't run/build unit tests in CI pipelines for platforms when there is not a worker of the appropriate model available. (!859)(1d5b13b7)
- **devops**: allow multiple devices of same platform in HWIL CI pipelines. (!862)(1332223f)
- **devops**: enable zcu102, picozed, microzed platforms in HWIL CI pipelines. (!862)(1332223f)
- **devops**: add source_sink application to CI/CD pipelines. (!867)(0eae0e52)
- **devops**: add testbias app to CI pipelines. (!875)(13b4d819)
- **devops**: prevent attempts to build inactive platforms in CI pipelines. (!875)(13b4d819)
- **devops**: let runner dictate device to connect to in HWIL CI pipelines. (!875)(13b4d819)
- **devops**: make pipeline ignore warnings from ocpidev. (!880)(ad9c8552)
- **devops**: add FSK app to CI pipelines. (!893)(feccd726)
- **devops**: add failed job report to CI pipelines. (!894)(c0026da7)
- **devops**: auto push develop and release docker images to gitlab container registry. (!954)(da29f8d2)
- **devops**: add zcu111 to pipeline. (!964)(7b6738bf)
- **devops**: fix section that pushes to gitlab registry to use predefined ci variables. (!985)(46482d35)
- **devops**: add x310 and x3u platforms to CI config to enable them to run properly in pipelines. (!1022)(243d0bbd)
- **devops**: update docs to support building docs in forked projects. (!1091)(80eba8b2)
- **devops,tools**: add `rocky8` RCC development platform support. (!874)(cd558c26)
- **devops,tools**: add `ubuntu22_04` RCC development platform support. (!885)(4f8b666c)
- **doc**: add install-opencpi man page. (!1021)(d179174a)
- **doc**: add ocpidev asset doc system usage information to relevant man pages. (!1023)(cb54577f)
- **doc**: fix ordering of options in man pages to be short form first. (!1023)(cb54577f)
- **doc**: add ethernet transport performance tuning guide. (!1033)(2f8d0d14)
- **doc**: rename HDL and RCC worker RST files to latest naming scheme. (!1035)(9d2a079f)
- **doc**: update component RST docs to use latest terminology and template format. (!1035)(9d2a079f)
- **doc**: update component RST block diagrams (*.odg, *.svg) to be exemplar. (!1035)(9d2a079f)
- **doc**: update RST files and directories for platform project device workers to use current organization, structure, and naming scheme. (!1036)(594fc672)
- **doc**: remove deprecated RST files and directories for platform project device workers. (!1036)(594fc672)
- **doc**: add node graph to Tutorial 1. (!1039)(123ce17f)
- **doc**: add node graph to tutorial 2. (!1040)(d85c577b)
- **doc**: move picoevb and zed RST files and images to be co-located with their assets. (!1041)(95806072)
- **doc**: update picoevb and zed GSG to refer to information in User Guide and Installation Guide. (!1041)(95806072)
- **doc**: add missing hyphen to `ocpidev` command for #3647. (!1142)(11896d00)
- **doc**: update tutorial docs for new code, and grammar issue #3646. (!1142)(11896d00)
- **doc,tools**: remove `--rcc-platform` from `ocpidev run` man page, for #3662. (!1142)(11896d00)
- **doc,tools**: add `ocpi_documentation_include` directive to ocpidoc. (!882)(877eb994)
- **doc,tools**: add RST templates for OSP-related documentation. (!1008)(82b22c6a)
- **hdl base**: add DG-RDMA Primitive Library. (!947)(7de0ace9)
- **hdl base**: add (optional) fairer arbitration scheme to SDP. (!947)(7de0ace9)
- **hdl base**: add DG-RDMA configuration device worker and proxy. (!947)(7de0ace9)
- **hdl base**: add Ethernet MAC dependency. (!947)(7de0ace9)
- **osp**: add Riviera-Pro support. (!906)(5e074965)
- **osp**: add zed_ether platform as an HDL DG-RDMA reference OSP. (!947)(7de0ace9)
- **runtime**: add ability to dump data when ocpirun receives CTL+C. (!1114)(6a5039d1)
- **runtime**: adjust error message and fix memory issue. (!872)(9fc49553)
- **runtime**: add worker property settings to system.xml, either app or artifact. (!877)(10c53c07)

tools/ocpidev2/ocpidev2.py
#!/usr/bin/python3

# This file is protected by Copyright. Please refer to the COPYRIGHT file
# distributed with this source distribution.
#
# This file is part of OpenCPI <http://www.opencpi.org>
#
# OpenCPI is free software: you can redistribute it and/or modify it under the
# terms of the GNU Lesser General Public License as published by the Free
# Software Foundation, either version 3 of the License, or (at your option) any
# later version.
#
# OpenCPI is distributed in the hope that it will be useful, but WITHOUT ANY
# WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR
# A PARTICULAR PURPOSE. See the GNU Lesser General Public License for more
# details.
#
# You should have received a copy of the GNU Lesser General Public License
# along with this program. If not, see <http://www.gnu.org/licenses/>.


""" This file exposes the Command Line Interface (CLI) below it, and above it
    interfaces with project-like things (ProjectDatabase), and not individual

    Authors
    =======
    Davis Hoover (git@davishoover.com)
    Joel Palmer (jpalmer@geontech.com)

    """

import os
import argparse
import signal
from _opencpi.assets.abstract2 import *
# IMPORTANT - interface with ProjectDatabase, nothing else
from _opencpi.assets.registry2 import ProjectDatabase


def ocpidevsignint(sig, frame):
    """ add create-specific arguments as per man ocpidev2-build """
    raise Exception('Ctrl-C stopped execution')


def add_build_create_show_arguments(parser):
    parser.add_argument('authoring_model', nargs='?', default='')
    parser.add_argument('noun', nargs='?', default=None)
    parser.add_argument('librarytype', nargs='?', default=None)
    parser.add_argument('name', nargs='?', default=None)
    return parser


def add_build_arguments(parser):
    """ add build-specific arguments as per the man
        page 'man ocpidev2-build' """
    parser.add_argument('--hdl-target', default=[], action='append')
    parser.add_argument('--hdl-platform', default=[], action='append')
    parser.add_argument('--rcc-platform', default=[], action='append')
    parser.add_argument('-j', nargs='?', type=int, default=1)
    parser = add_build_create_show_arguments(parser)
    return parser


def add_create_arguments(parser, noun, authoring_model=''):
    """ add create-specific arguments as per the man
        page 'man ocpidev2-create ' """
    parser.add_argument('-k', '--keep', default=False, action='store_true')
    asset = ProjectDatabase.get_asset_for_create(noun)
    if noun == 'application':
        group = parser.add_mutually_exclusive_group()
        group.add_argument('-X', '--xml-app', default=False,
                           action='store_true')
        group.add_argument('-x', '--xml-dir-app', default=False,
                           action='store_true')
    if noun in ['adapter', 'component', 'device', 'primitive', 'protocol',
                'worker', 'test']:
        group = parser.add_mutually_exclusive_group()
        group.add_argument('--hdl-library', nargs='?', default=None)
        group.add_argument('--library', default=None)
        parser.add_argument('-P', '--platform', default=False)
        if noun in ['component', 'protocol']:
            group.add_argument('-p', '--project', default=False,
                               action='store_true')
    if noun == 'component':
        parser.add_argument('-t', '--create-test', default=False,
                            action='store_true')
    if noun == 'project':
        parser.add_argument('--register', default=False, action='store_true')
    if noun == 'worker':
        parser.add_argument('-W', '--worker', default=[],
                            action='append')
    if asset is not None:
        for attr in asset.get_attr_infos():
            if attr.cli is not None:
                if attr.is_bool:
                    _default = [] if attr.is_list else False
                    _action = 'append' if attr.is_list else \
                              ('store_true' if attr.is_bool else 'store')
                    parser.add_argument(attr.cli[0], attr.cli[1],
                                        default=_default, action=_action)
                else:
                    _default = [] if attr.is_list else \
                               (0 if attr.is_int else '')
                    _action = 'append' if attr.is_list else \
                              ('store_true' if attr.is_bool else 'store')
                    parser.add_argument(attr.cli[0], attr.cli[1], nargs='?',
                                        default=_default, action=_action)
    parser = add_build_create_show_arguments(parser)
    return parser


def add_delete_arguments(parser):
    # TODO this should probably be done better
    return add_create_arguments(parser, '')


def add_show_arguments(parser):
    """ add create-specific arguments as per man ocpidev2-show """
    group1 = parser.add_mutually_exclusive_group()
    group1.add_argument('--local-scope', default=False, action='store_true')
    group1.add_argument('--global-scope', default=True, action='store_true')
    group = parser.add_mutually_exclusive_group()
    group.add_argument('--simple', default=False, action='store_true')
    group.add_argument('--table', default=False, action='store_true')
    group.add_argument('--json', default=False, action='store_true')
    parser = add_build_create_show_arguments(parser)
    return parser


def add_run_arguments(parser, noun):
    parser.add_argument('-G', '--only-platform', default=False,
                        action='store_true')
    parser.add_argument('-O', '--exclude-platform', default=False,
                        action='store_true')
    group = parser.add_mutually_exclusive_group()
    group.add_argument('--hdl-library', nargs='?', default='')
    group.add_argument('-M', '--library', default=None)
    parser.add_argument('--accumulate-errors', default=False,
                        action='store_true')
    parser.add_argument('--case', default=[], action='append')
    parser.add_argument('--keep-simulations', default=False,
                        action='store_true')
    parser.add_argument('--phase', default=[], action='append')
    parser.add_argument('--remotes', nargs='?', default='')
    parser.add_argument('--view', default=False, action='store_true')
    return parser


def get_cli_noun_verb_tuple(argv):
    verb = ''
    noun = ''
    state = 0
    verbs = ['build', 'clean', 'create', 'delete', 'show', 'run']
    singular_nouns = ['application', 'adapter', 'assembly', 'card',
                      'component', 'device', 'library', 'slot', 'test',
                      'primitive', 'project', 'protocol', 'platform',
                      'registry', 'worker']
    plural_nouns = ['applications', 'adapters', 'assemblies', 'components',
                    'devices', 'libraries', 'slots', 'tests', 'primitives',
                    'projects', 'platforms', 'protocols', 'workers']
    special_nouns = ['primitive', 'library', 'core']
    allow_exception = False
    # state machine with states:
    # 0: waiting on verb
    # 1: waiting on noun
    # 2: got a -<option> or --<option>, waiting on value of that option
    #    (necessary for intermixed arguments)
    for idx in range(len(argv)):
        # print(state)
        # print(argv[idx])
        if (state == 0) and (argv[idx] in verbs):
            verb = argv[idx]
            state = 1
        elif (state == 1):
            if argv[idx][0] == '-':
                state = 2
            elif argv[idx] in ['hdl', 'rcc', 'ocl']:
                pass  # authoring_model = argv[idx]
            elif argv[idx] in ['primitive', 'primitives']:
                noun = argv[idx]
                state = 2
            elif (argv[idx] == 'library') and (argv[idx-1] == 'primitive'):
                pass
            elif (argv[idx] == 'libraries') and (argv[idx-1] == 'primitive'):
                pass
            elif (argv[idx] == 'core') and (argv[idx-1] == 'primitive'):
                pass
            elif (argv[idx] == 'cores') and (argv[idx-1] == 'primitive'):
                pass
            elif argv[idx] in (singular_nouns + plural_nouns):
                noun = argv[idx]
                break
                state = 2
            elif allow_exception:
                raise Exception('invalid noun: ' + argv[idx])
        elif (state == 2):
            allow_exception = False
            state = 1
    return (noun, verb)


def get_arg_parser():
    parser = argparse.ArgumentParser(description='', add_help=False)
    parser.add_argument('-d', default=[], action='append')
    parser.add_argument('-h', '--help', action='store_true')
    parser.add_argument('-v', '--verbose', action='store_true')
    parser.add_argument('-l', '--log-level', nargs='?', default=0, type=int)
    # only intended to be used for tab completion
    parser.add_argument('--suppress-warn', action='store_true')
    parser.add_argument('verb', nargs='?', default='')
    (noun, verb) = get_cli_noun_verb_tuple(sys.argv)
    if verb == 'build':
        parser = add_build_arguments(parser)
    elif verb == 'create':
        parser = add_create_arguments(parser, noun)
    elif verb == 'show':
        parser = add_show_arguments(parser)
    elif verb == 'run':
        parser = add_run_arguments(parser)
    else:
        parser.add_argument('noun', nargs='?', default=None)
        parser.add_argument('name', nargs='?', default=None)
    return parser


def get_args(parser):
    # below 3 lines parse, allowing for posix conformance (intermixed args)
    (args, unknown_args) = parser.parse_known_args()
    for unknown_arg in unknown_args:
        if unknown_arg.startswith('application') or \
           unknown_arg.startswith('assembl') or \
           unknown_arg.startswith('component') or \
           unknown_arg.startswith('card') or \
           unknown_arg.startswith('device') or \
           unknown_arg.startswith('platform') or \
           unknown_arg.startswith('project') or \
           unknown_arg.startswith('protocol') or \
           unknown_arg.startswith('primitive') or \
           unknown_arg.startswith('registr') or \
           unknown_arg.startswith('slot') or \
           unknown_arg.startswith('librar') or \
           unknown_arg.startswith('target') or \
           unknown_arg.startswith('test') or \
           unknown_arg.startswith('worker'):
            args.noun = unknown_arg
        elif unknown_arg in get_authoring_models():
            args.auhoring_model = unknown_arg
        elif unknown_arg in ['core', 'cores', 'library', 'libraries']:
            args.noun = 'primitive'
            # later set to adjective...
            args.name = unknown_arg
        elif unknown_arg in ['--xml-dir-app']:
            args.xml_dir_app = unknown_arg  # no idea why this is necessary...
            args.xml_app = None  # no idea why this is necessary...
        else:
            raise Exception('invalid argument: ' + unknown_arg)
    return args


def get_cli_dict():
    """ get a dictionary of settings which looks like CLI
        cli_dict['d']                a list of entries corresponding to
                                     -d <dir>, with the list being empty if no
                                     -d was specified
        cli_dict['verb']             the CLI verb (always exists)
        cli_dict['noun']             the CLI noun (often exists)
        cli_dict['name']             the CLI noun (sometimes exists)
        cli_dict['authoringmodel']   either 'hdl' or 'rcc' or '' if no
                                     authoring model was specified
        cli_dict['adjective']        'core' or 'library' when
                                     'hdl primitive core' or
                                     'hdl primitive library' was specified at
                                     the command line (need to change this...)
        cli_dict[many..other..entries..] """
    args = get_args(get_arg_parser())
    nouns = ['adapter', 'adapters',
             'assembly', 'assemblies',
             'application', 'applications',
             'core', 'cores',
             'card', 'cards',
             'device', 'devices',
             'component', 'components',
             'platform', 'platforms',
             'primitive', 'primitives',
             'project', 'projects',
             'protocol', 'protocols',
             'registry',
             'library', 'libraries',
             'slot', 'slots',
             'target', 'targets',
             'test', 'tests',
             'worker', 'workers']
    # Logger().debug('args : ' + str(args))
    try:
        args.adjective = ''
        if args.authoring_model in nouns:
            # first correction - extract proper authoring model and align the
            # rest
            args.name = args.noun
            args.noun = args.authoring_model
            args.authoring_model = ''
        if args.name is None:
            if args.librarytype != '':
                args.name = args.librarytype
                del args.librarytype
        # Logger().debug('args : ' + str(args))
        if (args.name in nouns) and args.noun.startswith('primitive'):
            if args.name == 'core':
                args.adjective = 'core'
                args.noun = 'primitive'
            elif args.name == 'cores':
                args.adjective = 'core'
                args.noun = 'primitives'
            elif args.name == 'library':
                args.adjective = 'library'
                args.noun = 'primitive'
            elif args.name == 'libraries':
                args.adjective = 'library'
                args.noun = 'primitives'
            elif (args.name != 'platform') and (args.name != 'protocol'):
                args.adjective = ''
                args.noun = args.name
            if (args.name != 'platform') and (args.name != 'protocol'):
                args.name = None
        # Logger().debug('args : ' + str(args))
    except AttributeError:
        # TODO replace this hack
        args.authoring_model = ''
    # TODO move below 3 lines to AssetBase once proper checks in place
    if args.verb != 'unittest':
        if args.noun is not None:
            if not ProjectDatabase.is_worker(args.noun):
                if args.name:
                    if args.noun.startswith('test'):
                        if '.test' in args.name:
                            args.name = args.name.split('.test')[0]
        if not args.help:
            if (args.noun is not None) and (args.noun not in nouns):
                if args.verb != 'apply':
                    msg = 'noun ' + str(args.noun) + ' is not supported'
                    raise Exception(msg)
    tmp = args
    mylist = args.d.copy()
    tmp.d = []
    for _dir in mylist:
        _dir = os.path.normpath(os.path.abspath(_dir))
        if _dir in tmp.d:
            Logger().warn('-d has duplicate ' + _dir)
        else:
            tmp.d.append(_dir)
    cli_dict = vars(tmp)
    # make CLI look like attrs (necessary for create cli verb)
    cli_dict = ({key.replace('_', ''): val for key, val in cli_dict.items()})
    if (cli_dict['noun'] == 'worker') or (cli_dict['noun'] == 'device') or \
       (cli_dict['noun'] == 'adapter'):
        if cli_dict['name'] is None:
            msg = 'name must be specified for \'' + cli_dict['noun'] + '\''
            raise Exception(msg)
        authoring_model_from_name = cli_dict['name'].split('.')[-1]
        authoring_model_is_in_name = '.' in cli_dict['name']
        if authoring_model_is_in_name:
            if authoring_model_from_name not in ['hdl', 'rcc']:
                msg = '\'' + authoring_model_from_name + '\' is an '
                msg += 'invalid authoring model (must be \'hdl\' or '
                msg += '\'rcc\')'
                raise Exception(msg)
        if cli_dict['authoringmodel'] == '':
            if not authoring_model_is_in_name:
                msg = cli_dict['name']
                msg += ' does not have an authoring model, please format the '
                msg += 'name as <foo>.<model>, e.g., \'ocpidev2 create '
                msg += 'worker foo.hdl\''
                raise Exception(msg)
            cli_dict['authoringmodel'] = authoring_model_from_name
        else:
            if cli_dict['authoringmodel'] not in ['hdl', 'rcc']:
                msg = '\'' + cli_dict['authoringmodel'] + '\' is an '
                msg += 'invalid authoring model (must be \'hdl\' or '
                msg += '\'rcc\')'
                raise Exception(msg)
            if authoring_model_is_in_name:
                if cli_dict['authoringmodel'] != authoring_model_from_name:
                    msg = 'authoring model '
                    msg += '\'' + cli_dict['authoringmodel'] + '\' '
                    msg += 'does not match authoring model specified in name '
                    msg += '\'' + cli_dict['name'] + '\' '
                    raise Exception(msg)
        if not authoring_model_is_in_name:
            msg = '<name> ' + '\'' + cli_dict['name'] + '\''
            msg += ' was not specified with an authoring model suffix, '
            msg += 'it is recommended to '
            msg += 'format the name as <foo>.<model>, e.g., \'ocpidev2 '
            msg += 'create worker foo.hdl\''
            Logger().warn(msg)
        cli_dict['name'] = cli_dict['name'].split('.')[0]
    if cli_dict['noun'] is not None:
        bad_hdl = (cli_dict['authoringmodel'] == 'hdl') and \
                  (not ProjectDatabase.is_hdl(cli_dict['noun']))
        bad_rcc = (cli_dict['authoringmodel'] == 'rcc') and \
                  (not ProjectDatabase.is_rcc(cli_dict['noun']))
        msg = 'authoring model '
        if bad_hdl or bad_rcc:
            if bad_none:
                msg += 'must be specified '
            else:
                msg += '\'' + cli_dict['authoringmodel'] + '\' '
                msg += 'is invalid '
            msg += 'for noun \'' + cli_dict['noun'] + '\''
            raise Exception(msg)
    if cli_dict['noun'] is not None:
        if cli_dict['noun'].endswith('s'):
            if cli_dict['name'] is not None:
                msg = 'cannot specify name when noun \'' + cli_dict['noun']
                msg += '\' is plural'
                raise Exception(msg)
        else:
            if ((cli_dict['name'] is None) and (cli_dict['d'] == [])) and \
               (cli_dict['noun'] != 'registry'):
                if cli_dict['verb'] != 'register':
                    raise Exception('must specify name or -d')
    if ('worker' in cli_dict.keys()) and (cli_dict['worker'] != []):
        raise Exception('do not use --worker, instead create separate workers')
    if 'localscope' not in cli_dict.keys():
        cli_dict['localscope'] = False
    if 'globalscope' not in cli_dict.keys():
        cli_dict['globalscope'] = True
    if cli_dict['name'] is not None:
        if cli_dict['name'] in ['gen', 'lib']:
            raise Exception('name \'' + cli_dict['name'] + '\' is not allowed')
    return cli_dict


# TODO delete
global g_suppress_warn


def main():
    ret = 0
    try:
        signal.signal(signal.SIGINT, ocpidevsignint)
        cli_dict = get_cli_dict()
        if cli_dict['suppresswarn']:
            set_g_suppress_warn(True)
        if cli_dict['loglevel']:
            set_g_log_level(cli_dict['loglevel'])
        Logger().debug('cli_dict : ' + str(cli_dict))
        ProjectDatabase().dispatch_verb(cli_dict)
    except Exception as exception:
        Logger().error(str(exception))
        ret = 1
    return ret


if __name__ == '__main__':
    exit(main())
